### PR TITLE
Wallet: Remove global coinSelector and allowSpendingUnconfirmedTransactions().

### PR DESCRIPTION
**Pull request by schildbach**
Coin selection is a per SendRequest, per createSend() or per getBalance() call affair.
Having it wallet-global can lead to race conditions, as sometimes experienced in the
WalletTest unit tests.

### CODE REVIEW

**Comment by msgilligan** [comment]
@schildbach Does this fix (some/all) the unit test intermittent failures?

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java [hunk 1]
```diff
@@ -230,8 +230,6 @@ protected boolean removeEldestEntry(Map.Entry<Sha256Hash, Transaction> eldest) {
     // that was created after it. Useful when you believe some keys have been compromised.
     private volatile long vKeyRotationTimestamp;
 
-    protected CoinSelector coinSelector = DefaultCoinSelector.get();
```
Comment (msgilligan) [inline core/src/main/java/org/bitcoinj/wallet/Wallet.java]: I would consider leaving this and making it `final` -- this will allow subclasses and tests more flexibility in overriding.

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java [hunk 2]
```diff
@@ -3710,7 +3708,7 @@ public Coin getBalance(BalanceType balanceType) {
         try {
             if (balanceType == BalanceType.AVAILABLE || balanceType == BalanceType.AVAILABLE_SPENDABLE) {
                 List<TransactionOutput> candidates = calculateAllSpendCandidates(true, balanceType == BalanceType.AVAILABLE_SPENDABLE);
-                CoinSelection selection = coinSelector.select(NetworkParameters.MAX_MONEY, candidates);
+                CoinSelection selection = DefaultCoinSelector.get().select(NetworkParameters.MAX_MONEY, candidates);
```
Comment (msgilligan) [inline core/src/main/java/org/bitcoinj/wallet/Wallet.java]: I would consider having a `protected final` default coin selector to allow subclasses to override, rather than hardcoding `DefaultCoinSelector`.

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java [hunk 3]
```diff
@@ -3952,9 +3950,49 @@ public Coin getTotalSent() {
...
+     * @throws MultipleOpReturnRequested if there is more than one OP_RETURN output for the resultant transaction.
+     * @throws BadWalletEncryptionKeyException if the supplied {@link SendRequest#aesKey} is wrong.
+     */
+    public Transaction createSend(Address address, Coin value, boolean allowUnconfirmed)
+            throws InsufficientMoneyException, BadWalletEncryptionKeyException {
         SendRequest req = SendRequest.to(address, value);
         if (params.getId().equals(NetworkParameters.ID_UNITTESTNET))
             req.shuffleOutputs = false;
+        if (allowUnconfirmed)
+            req.allowUnconfirmed();
```
Comment (msgilligan) [inline core/src/main/java/org/bitcoinj/wallet/Wallet.java]: I would try to make SendRequest more immutable and move the setting of the coin selector to the constructor, but that could be a separate PR.

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java [hunk 4]
```diff
@@ -4471,40 +4509,6 @@ public boolean canSignFor(Script script) {
         return candidates;
     }
 
-    /** Returns the {@link CoinSelector} object which controls which outputs can be spent by this wallet. */
-    public CoinSelector getCoinSelector() {
```
Comment (msgilligan) [inline core/src/main/java/org/bitcoinj/wallet/Wallet.java]: I might leave this method or make it `getDefaultCoinSelector` if you take my advice on leaving it in a `protected final` field.

**Comment by schildbach** [comment]
> @schildbach Does this fix (some/all of) the unit test intermittent failures?

I guess it could fix some. Some tests changed the coinSelector and didn't change it back.

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java (hunk 1 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/wallet/Wallet.java]: That's a good idea. I'm leaving it in.

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java (hunk 2 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/wallet/Wallet.java]: Reverted this line and the similar ones.

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java (hunk 3 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/wallet/Wallet.java]: Yes, that's another PR. I always wanted it immutable (and return the results in a separate object).

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java (hunk 4 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/wallet/Wallet.java]: I'm leaving it in, and changed the JavaDoc a bit.

**Comment by schildbach** [comment]
I addressed your comments.

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java [hunk 5]
```diff
@@ -4471,7 +4511,7 @@ public boolean canSignFor(Script script) {
         return candidates;
     }
 
-    /** Returns the {@link CoinSelector} object which controls which outputs can be spent by this wallet. */
+    /** Returns the default {@link CoinSelector} object that is used by this wallet if no custom selector is specified. */
```
Comment (msgilligan) [inline core/src/main/java/org/bitcoinj/wallet/Wallet.java]: Comment update looks great!
